pkgs/benchmark_harness/lib/src/bench_command/compile_and_run.dart

```diff
@@ -0,0 +1,186 @@
+    );
+  }
+
+  for (var mode in options.flavor) {
+    await _Runner(flavor: mode, target: options.target).run();
+  }
+}
+
+class BenchException implements Exception {
+  const BenchException(this.message, this.exitCode) : assert(exitCode > 0);
+  final String message;
+  final int exitCode;
+
+  @override
+  String toString() => 'BenchException: $message ($exitCode)';
+}
+
+/// Base name for output files.
+const _outputFileRoot = 'out';
+
+/// Denote the "stage" of the compile/run step for logging.
+enum _Stage { compile, run }
+
+/// Base class for runtime-specific runners.
+abstract class _Runner {
+  _Runner._({required this.target, required this.flavor})
+      : assert(FileSystemEntity.isFileSync(target), '$target is not a file');
+
+  factory _Runner({required RuntimeFlavor flavor, required String target}) {
+    return (switch (flavor) {
+      RuntimeFlavor.jit => _JITRunner.new,
+      RuntimeFlavor.aot => _AOTRunner.new,
+      RuntimeFlavor.js => _JSRunner.new,
+      RuntimeFlavor.wasm => _WasmRunner.new,
+    })(target: target);
+  }
+
+  final String target;
+  final RuntimeFlavor flavor;
+  late Directory _tempDirectory;
+
+  /// Executes the compile and run cycle.
+  ///
+  /// Takes care of creating and deleting the corresponding temp directory.
+  Future<void> run() async {
+    _tempDirectory = Directory.systemTemp
+        .createTempSync('bench_${DateTime.now().millisecondsSinceEpoch}_');
+    try {
+      await _runImpl();
+    } finally {
+      _tempDirectory.deleteSync(recursive: true);
+    }
+  }
+
+  /// Overridden in implementations to handle the compile and run cycle.
+  Future<void> _runImpl();
+
+  /// Executes the specific [executable] with the provided [args].
+  ///
+  /// Also prints out a nice message before execution denoting the [flavor] and
+  /// the [stage].
+  Future<void> _runProc(
+      _Stage stage, String executable, List<String> args) async {
+    print('''
+\n${flavor.name.toUpperCase()} - ${stage.name.toUpperCase()}
+$executable ${args.join(' ')}
+''');
+
+    final proc = await Process.start(executable, args,
+        mode: ProcessStartMode.inheritStdio);
+
+    final exitCode = await proc.exitCode;
+
+    if (exitCode != 0) {
+      throw ProcessException(executable, args, 'Process errored', exitCode);
+    }
+  }
+
+  String _outputFile(String ext) =>
+      _tempDirectory.uri.resolve('$_outputFileRoot.$ext').toFilePath();
+}
+
+class _JITRunner extends _Runner {
+  _JITRunner({required super.target}) : super._(flavor: RuntimeFlavor.jit);
+
+  @override
+  Future<void> _runImpl() async {
+    await _runProc(_Stage.run, Platform.executable, [target]);
+  }
+}
+
+class _AOTRunner extends _Runner {
+  _AOTRunner({required super.target}) : super._(flavor: RuntimeFlavor.aot);
+
+  @override
+  Future<void> _runImpl() async {
+    final outFile = _outputFile('exe');
+    await _runProc(_Stage.compile, Platform.executable, [
+      'compile',
```

Comment:
In the future, this should probably also support `dart build` for packages using native assets.